mahiru/components/domain_administrator.py

```diff
@@ -79,142 +110,162 @@ def execute_step(
+                        asset.image_location, image_file)
 
-        return image_files
+                with image_file.open('rb') as f:
+                    image = self._dcli.images.load(f.read())[0]
 
-    def _load_images_into_docker(
-            self, job_id: int, image_files: Dict[str, Path]
-            ) -> Dict[str, Image]:
-        """Load images from files into Docker.
+                self._loaded_images[asset.id] = image.id
+                self._loaded_images_refs[asset.id] = 1
+                image_file.unlink()
 
-        This loads container images from tarballs on disk into the
-        local Docker image repository.
+            return image
+
+    def _free_image(self, asset_id: str) -> None:
+        """Decrements the use count on the asset image.
+
+        If this was the last user, the image is removed from Docker,
+        and will be re-downloaded next time it is needed.
 
         Args:
-            job_id: Id of the step execution job these are for.
-            image_files: Files (tarballs) to load in.
+            asset_id: The id of the asset that's no longer needed by
+                    the caller.
+        """
+        with self._loaded_images_lock:
+            if asset_id not in self._loaded_images_refs:
+                raise KeyError('Asset not found.')
+
+            self._loaded_images_refs[asset_id] -= 1
+            if self._loaded_images_refs[asset_id] == 0:
+                image_id = self._loaded_images[asset_id]
+                try:
+                    self._dcli.images.remove(image_id)
+                except docker.errors.ImageNotFound:
+                    # Base output images may already have been deleted,
+                    # so this is probably okay.
+                    pass
+                except Exception as e:
+                    logger.warning(f'Failed to remove image {image_id}: {e}')
+                    self._dcli.images.remove(image_id, force=True)
+                finally:
+                    del self._loaded_images[asset_id]
+
+    def _ensure_images_available(
+            self, workdir: Path, assets: Dict[str, Asset],
+            ) -> Dict[str, Image]:
+        """Ensure required images are available in Docker.
+
+        Args:
+            workdir: Working directory for this job.
+            assets: Asset objects to download.
 
         Returns:
             Docker Image objects indexed by input/output name.
         """
+        images = dict()
+
         try:
-            images = dict()
-            for name, path in image_files.items():
-                with path.open('rb') as f:
-                    images[name] = self._dcli.images.load(f.read())[0]
-            return images
+            for name, asset in assets.items():
+                images[name] = self._ensure_image_available(asset, workdir)
+
         except Exception:
-            for image in images.values():
-                self._dcli.images.remove(image.id)
+            for name in images:
```

Comment:
Can you add a comment here to explain why it makes sense (which it probably does) to free all images in case of an exception. (i.e. in case any of the images is not there, we free all images).

Reply:
If any of the images cannot be made available, then we undo all the work we've done already before we raise the exception so that the Docker daemon is either in the desired state (on success) or in its previous state (on error). This is known as strong exception safety in the C++ world, and it's important to ensure that the system keeps running correctly even if errors occur. I'm not making any such promises on that about this Mahiru implementation, but at least this function gets it right :-). (Weak exception safety, in case you were wondering, leaves the system/object in some well-defined/consistent state, but not necessarily the original one.)

I'll add a brief comment.